### DISCUSSION TITLE
Many nested load options drastically slow down performance (despite underlying queries being fast)

### DISCUSSION THREAD

**Discussion by nithiwatter** [sqlalchemy/sqlalchemy#11246] [Usage Questions] [answered]
Hi,

We are currently debugging long-running endpoints which only happen when we use ORM-style queries with a lot of load options. We have tried profiling and believe that having a lot of load options ex.

```
stmt = select(A).options(*lot_of_options)
```

is detrimental to performance (aka - the time taken to perform JOINs via `joinedload` and separate queries via `selectinload` at the DBAPI level, psycopg, is way faster than SQLAlchemy processing the fetched rows from the DB).

Our load options are extremely nested (at least 200+ in total), and here are our profiled code:

```
         2662542 function calls (653845 primitive calls) in 2.470 seconds

   Ordered by: internal time
   List reduced from 1270 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
17754/330    0.760    0.000    1.292    0.004 /usr/local/lib/python3.10/site-packages/sqlalchemy/orm/strategy_options.py:810(<listcomp>)
1909578/330    0.447    0.000    1.295    0.004 /usr/local/lib/python3.10/site-packages/sqlalchemy/orm/strategy_options.py:796(_deep_clone)
       39    0.409    0.010    0.411    0.011 {method 'execute' of 'psycopg2.extensions.cursor' objects}
20592/546    0.093    0.000    0.290    0.001 /usr/local/lib/python3.10/site-packages/sqlalchemy/orm/strategy_options.py:756(<genexpr>)
10491/195    0.054    0.000    0.294    0.002 /usr/local/lib/python3.10/site-packages/sqlalchemy/orm/strategy_options.py:704(_gen_cache_key)
 33178/39    0.049    0.000    0.309    0.008 /usr/local/lib/python3.10/site-packages/sqlalchemy/sql/traversals.py:159(_gen_cache_key)
    43173    0.041    0.000    0.110    0.000 /usr/local/lib/python3.10/site-packages/sqlalchemy/sql/traversals.py:525(<genexpr>)
    18152    0.032    0.000    0.082    0.000 /usr/local/lib/python3.10/site-packages/sqlalchemy/orm/strategy_options.py:232(_generate)
    10491    0.023    0.000    0.134    0.000 /usr/local/lib/python3.10/site-packages/sqlalchemy/sql/traversals.py:522(visit_multi_list)
    18351    0.023    0.000    0.050    0.000 /usr/local/lib/python3.10/site-packages/sqlalchemy/sql/base.py:557(_generate)
```

I might be interpreting this wrong but it seems like the `execute` via `psycopg` is only around 0.41s, whereas the list comprehension inherent via `_adapt_cached_option_to_uncached_option` (which seems to always be called even if our queries were already cached) is taking almost 0.8s.

We have confirmed that all these queries were cached via engine logging - so this is not slowness due to the typical cache misses during compilation.

We are on 1.4.46 and would appreciate any advice on how to improve performance here - willing to provide anymore info as necessary for further debugging. Thank you 😊

**Comment by zzzeek** [comment]
Hi -

there was a major performance fix for this function in 2.0.28 listed at https://docs.sqlalchemy.org/en/20/changelog/changelog_20.html#change-ac3f7e555f345c82b2b07b34069c7489 , however that was mostly to fix a performance regression here that was introduced in 2.0.23.  nevertheless, the newer approach is all around more efficient and while you are on 1.4.46 it would be very helpful if you could at least test that codepath on 2.0.23 to see if there's improvement.

**Reply by zzzeek** [reply]
otherwise, you can skip the cache key entirely for this query, which might fix the problem for now, using execution_options(query_cache_size=0) on your execute() call.

**Reply by nithiwatter** [reply]
the above option did nothing - after profiling with no cache - we are still calling list comprehension/deep clone the same amount of times as with caching (I suspect that list comprehension + deep clone logic is always called whether or not the query is cached).

For further debugging, I have tried logging `_extra_criteria` for all the 200+ load options and all of them are tuples with 0 length, so I am very confused why `_deep_clone` would take long given that

```
    def _adjust_for_extra_criteria(self, context):
        """Apply the current bound parameters in a QueryContext to all
        occurrences "extra_criteria" stored within al this Load object;
        copying in place.

        """
        orig_query = context.compile_state.select_statement

        applied = {}

        ck = [None, None]

        def process(opt):
            if not opt._extra_criteria:
                return

            if ck[0] is None:
                ck[:] = (
                    orig_query._generate_cache_key(),
                    context.query._generate_cache_key(),
                )
            k1, k2 = ck

            opt._extra_criteria = tuple(
                k2._apply_params_to_element(k1, crit)
                for crit in opt._extra_criteria
            )

        return self._deep_clone(applied, process)
```

We should short-circuit in `process`? The above code is from 1.4.46 so might be a different than the one mentioned in this issue - https://github.com/sqlalchemy/sqlalchemy/discussions/11246

**Reply by zzzeek** [reply]
oh in fact, I also forgot - the structure of options in 1.4 was completely replaced in 2.0 with a vastly simpler structure, so there is no "deep_clone" operation like that anymore .    

so I would again ask that you please test with SQLAlchemy 2.0 to see if this resolves the issue

**Reply by zzzeek** [reply, marked as answer]
1.4 has a structure called _UnboundLoad that has a nested structure here, the _deep_clone in your perf output is that function.   _UnboundLoad is totally removed in 2.0 which was a major performance enhancement

**Reply by nithiwatter** [reply]
Sorry for the last question - but assuming that the migration process from 1.4 to 2.0 will take a while - is there any bandaid/hot fix we can do in the meanwhile? Given that the uncached option does not work

**Reply by zzzeek** [reply]
sure, just patch that whole routine and dont use the extra_criteria feature

Q would be why was this perf issue noticed suddenly?

**Reply by nithiwatter** [reply]
Kinda dumb but originally we thought our SQLs were slow, so we have done a lot of optimizing/improvements at the DB/SQL levels - but then we started noticing this instead 😓

**Reply by zzzeek** [reply]
I'm not seeing that we have any perf tests on this end that actually tried to measure a change by removing this method, if you have any sample programs that illustrate your performance issue that would be useful to compare under 2.0
